tests/components/tasmota/test_discovery.py

```diff
@@ -0,0 +1,416 @@
+"""The tests for the MQTT discovery."""
+import copy
+import json
+
+from hatasmota.const import CONF_ONLINE
+import pytest
+
+from homeassistant.components.tasmota.const import DEFAULT_PREFIX
+from homeassistant.components.tasmota.discovery import ALREADY_DISCOVERED
+
+from .conftest import setup_tasmota
+
+from tests.async_mock import AsyncMock, patch
+from tests.common import async_fire_mqtt_message
+
+DEFAULT_CONFIG = {
+    "dn": "My Device",
+    "fn": ["Beer", "Milk", "Three", "Four", "Five"],
+    "hn": "tasmota_49A3BC",
+    "mac": "00000049A3BC",
+    "md": "Sonoff 123",
+    "ofl": "offline",
+    CONF_ONLINE: "online",
+    "state": ["OFF", "ON", "TOGGLE", "HOLD"],
+    "sw": "2.3.3.4",
+    "t": "tasmota_49A3BC",
+    "t_f": "%topic%/%prefix%/",
+    "t_p": ["cmnd", "stat", "tele"],
+    "li": [0, 0, 0, 0, 0, 0, 0, 0],
+    "rl": [0, 0, 0, 0, 0, 0, 0, 0],
+    "se": [],
+    "ver": 1,
+}
+
+
+async def test_subscribing_config_topic(hass, mqtt_mock):
+    """Test setting up discovery."""
+    await setup_tasmota(hass)
+
+    discovery_topic = DEFAULT_PREFIX
+
+    assert mqtt_mock.async_subscribe.called
+    call_args = mqtt_mock.async_subscribe.mock_calls[0][1]
+    assert call_args[0] == discovery_topic + "/#"
+    assert call_args[2] == 0
+
+
+async def test_invalid_topic(hass, mqtt_mock):
+    """Test receiving discovery message on wrong topic."""
+    await setup_tasmota(hass)
+    with patch(
+        "homeassistant.components.tasmota.discovery.async_dispatcher_send"
```

Comment:
Patch the tasmota library target instead.

Reply:
In real I want to check that the closure `async_device_discovered` here is not called:
https://github.com/home-assistant/core/blob/74e4b35f17ec2cd18ae843edeccbfab9de3521cf/homeassistant/components/tasmota/discovery.py#L48

Fixed by making `async_device_discovered` a non inner function and patching it instead.

Reply:
We should not patch our own code. Patch the library.